crates/biome_js_analyze/src/lint/nursery/no_import_cycles.rs

```diff
@@ -0,0 +1,185 @@
+    /// Prevent import cycles.
+    ///
+    /// This rule warns when a file imports another file that, either directly
+    /// or indirectly, imports the original file again.
+    ///
+    /// Cycles can lead to symbols that are unexpectedly `undefined` and are
+    /// generally considered poor code hygiene.
+    ///
+    /// If a cycle is detected, it is advised to move code such that imports
+    /// only go in a single direction, i.e. they don't point "back" to the
+    /// importing file.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// **`foobar.js`**
+    /// ```js
+    /// import { baz } from "./baz.js";
+    ///
+    /// export function foo() {
+    ///     baz();
+    /// }
+    ///
+    /// export function bar() {
+    ///     console.log("foobar");
+    /// }
+    /// ```
+    ///
+    /// **`baz.js`**
+    /// ```js
+    /// import { bar } from "./foobar.js";
+    ///
+    /// export function baz() {
+    ///     bar();
+    /// }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// **`foo.js`**
+    /// ```js
+    /// import { baz } from "./baz.js";
+    ///
+    /// export function foo() {
+    ///     baz();
+    /// }
+    /// ```
+    ///
+    /// **`bar.js`**
+    /// ```js
+    /// export function bar() {
+    ///     console.log("foobar");
+    /// }
+    /// ```
+    ///
+    /// **`baz.js`**
+    /// ```js
+    /// import { bar } from "./bar.js";
+    ///
+    /// export function baz() {
+    ///     bar();
+    /// }
+    /// ```
+    ///
+    pub NoImportCycles {
+        version: "next",
+        name: "noImportCycles",
+        language: "js",
+        sources: &[
+            RuleSource::EslintImport("no-cycle"),
+        ],
+        recommended: false,
+    }
+}
+
+impl Rule for NoImportCycles {
+    type Query = ResolvedImports;
+    type State = Vec<String>;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Self::Signals {
+        let file_imports = ctx.imports_for_path(ctx.file_path())?;
+
+        let name_token = ctx.query().module_name_token()?;
+        let specifier_text = inner_string_text(&name_token);
+        let specifier = specifier_text.text();
+        let import = file_imports
+            .static_imports
+            .get(specifier)
+            .or_else(|| file_imports.dynamic_imports.get(specifier))?;
+        let resolved_path = import.resolved_path.as_ref().ok()?;
+
+        let imports = ctx.imports_for_path(resolved_path)?;
+        find_cycle(
+            ctx,
+            imports,
+            &mut BTreeSet::new(),
+            vec![resolved_path.to_string()],
+        )
+    }
+
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        let node = ctx.query();
+
+        let cwd = Utf8PathBuf::from(
+            std::env::current_dir()
+                .map(|cwd| cwd.to_string_lossy().to_string())
+                .unwrap_or_default(),
```

Comment:
I'm not sure this is correct, because the current working directory is now where the `biome.json` is found. Plus, `current_dir` might differ between CLI and LSP.

The use of `ctx.file_path()` might help here.

Reply:
Yeah, I had some doubts myself what would be best here, but I think using something relative to the `biome.json` actually makes sense. After all, we want to strip a common base from the paths, in such a way that it is easy to recognize where they are. The only motivation is that using the full absolute path is just unnecessarily noisy for human readers.

`ctx.file_path()` doesn't really help with that, since the paths that get imported may be higher up in the hierarchy than the one from where we start, so what would the common base be then?

Reply:
Anyway, we could actually create an ad-hoc diagnostic/type that leverage the `Location` type, and it will resolve the paths once it's rendered at console (via `biome_console::fmt::Display`) by passing the working directory.